### ISSUE TITLE
Fix URI encoding problems

### ISSUE THREAD

**Issue by m1ci** [rdfhdt/hdt-cpp#12]
I've converted nt file to HDT and loaded the HDT for querying in Jena. However, some resources are not queryable although they are present in the dataset.

The dataset is

```
<http://ru.dbpedia.org/resource/Список_римско-католических_епархий_(структурный_вид)> <http://dbpedia.org/ontology/wikiPageWikiLink> <http://ru.dbpedia.org/resource/Епархия_Абаэтетубы> .
<http://ru.dbpedia.org/resource/Епархия_Абаэтетубы> <http://dbpedia.org/ontology/wikiPageWikiLink> <http://ru.dbpedia.org/resource/Абаэтетуба> .
<http://ru.dbpedia.org/resource/Епархия_Абаэтетубы> <http://dbpedia.org/ontology/wikiPageWikiLink> <http://ru.dbpedia.org/resource/Бразилия> .
<http://ru.dbpedia.org/resource/Епархия_Абаэтетубы> <http://dbpedia.org/ontology/wikiPageWikiLink> <http://ru.dbpedia.org/resource/25_ноября> .
<http://ru.dbpedia.org/resource/Епархия_Абаэтетубы> <http://dbpedia.org/ontology/wikiPageWikiLink> <http://ru.dbpedia.org/resource/1961_год> .
<http://ru.dbpedia.org/resource/Епархия_Абаэтетубы> <http://dbpedia.org/ontology/wikiPageWikiLink> <http://ru.dbpedia.org/resource/Собор_Непорочного_зачатия_Пресвятой_Девы_Марии_(Абаэтетуба)> .
<http://ru.dbpedia.org/resource/Епархия_Абаэтетубы> <http://dbpedia.org/ontology/wikiPageWikiLink> <http://ru.dbpedia.org/resource/Архиепархия_Белен-до-Пара> .
<http://ru.dbpedia.org/resource/Епархия_Абаэтетубы> <http://dbpedia.org/ontology/wikiPageWikiLink> <http://ru.dbpedia.org/resource/Епархия> .
<http://ru.dbpedia.org/resource/Епархия_Абаэтетубы> <http://dbpedia.org/ontology/wikiPageWikiLink> <http://ru.dbpedia.org/resource/Римско-католическая_церковь> .
<http://ru.dbpedia.org/resource/Епархия_Абаэтетубы> <http://dbpedia.org/ontology/wikiPageWikiLink> <http://ru.dbpedia.org/resource/Кафедра_(христианство)> .
<http://ru.dbpedia.org/resource/Епархия_Абаэтетубы> <http://dbpedia.org/ontology/wikiPageWikiLink> <http://ru.dbpedia.org/resource/Собор_(храм)> .
<http://ru.dbpedia.org/resource/Епархия_Абаэтетубы> <http://dbpedia.org/ontology/wikiPageWikiLink> <http://ru.dbpedia.org/resource/Папство> .
<http://ru.dbpedia.org/resource/Епархия_Абаэтетубы> <http://dbpedia.org/ontology/wikiPageWikiLink> <http://ru.dbpedia.org/resource/Иоанн_XXIII> .
<http://ru.dbpedia.org/resource/Епархия_Абаэтетубы> <http://dbpedia.org/ontology/wikiPageWikiLink> <http://ru.dbpedia.org/resource/Булла> .
<http://ru.dbpedia.org/resource/Епархия_Абаэтетубы> <http://dbpedia.org/ontology/wikiPageWikiLink> <http://ru.dbpedia.org/resource/Территориальное_аббатство> .
<http://ru.dbpedia.org/resource/Епархия_Абаэтетубы> <http://dbpedia.org/ontology/wikiPageWikiLink> <http://ru.dbpedia.org/resource/4_августа> .
<http://ru.dbpedia.org/resource/Епархия_Абаэтетубы> <http://dbpedia.org/ontology/wikiPageWikiLink> <http://ru.dbpedia.org/resource/1981_год> .
<http://ru.dbpedia.org/resource/Епархия_Абаэтетубы> <http://dbpedia.org/ontology/wikiPageWikiLink> <http://ru.dbpedia.org/resource/Иоанн_Павел_II> .
<http://ru.dbpedia.org/resource/Епархия_Абаэтетубы> <http://dbpedia.org/ontology/wikiPageWikiLink> <http://ru.dbpedia.org/resource/Священник> .
<http://ru.dbpedia.org/resource/Епархия_Абаэтетубы> <http://dbpedia.org/ontology/wikiPageWikiLink> <http://ru.dbpedia.org/resource/Annuario_Pontificio> .
<http://ru.dbpedia.org/resource/Епархия_Абаэтетубы> <http://dbpedia.org/ontology/wikiPageWikiLink> <http://ru.dbpedia.org/resource/Acta_Apostolicae_Sedis> .
<http://ru.dbpedia.org/resource/Епархия_Абаэтетубы> <http://dbpedia.org/ontology/wikiPageWikiLink> <http://ru.dbpedia.org/resource/Категория:Католические_епархии_Бразилии> .
<http://ru.dbpedia.org/resource/Архиепархия_Белен-до-Пара> <http://dbpedia.org/ontology/wikiPageWikiLink> <http://ru.dbpedia.org/resource/Епархия_Абаэтетубы> .
<http://ru.dbpedia.org/resource/Епархия_Абаэтетубы> <http://www.w3.org/2000/01/rdf-schema#label> "Епархия Абаэтетубы"@ru .
```

In Jena the following query does not match anything:

```
StmtIterator tmpIter = model.listStatements(model.getResource("http://ru.dbpedia.org/resource/Епархия_Абаэтетубы"), null, (RDFNode)null);
if(!tmpIter.hasNext())
    System.out.println("No statements for this resource");
```

Loading the same dataset in Jena but as a ttl file works.

Maybe a bug in the HDT conversion process? Note that we are dealing with Russian cyrillic chars.

**Comment by hanito** [comment]
Di you manage to compile the code? m1ci

**Comment by m1ci** [comment]
yes, I use the hdt-cpp lib.

**Comment by mielvds** [comment]
yeah, HDT has encoding issues. Are you using UTF-8?

**Comment by RubenVerborgh** [comment]
Have you used SERD and `-f turtle` to create the HDT file? All other parsers are broken.

**Comment by m1ci** [comment]
> yeah, HDT has encoding issues. Are you using UTF-8?

I use subject and the object of the triples are IRIs not URIs.

`<http://ru.dbpedia.org/resource/Список_римско-католических_епархий_(структурный_вид)> <http://dbpedia.org/ontology/wikiPageWikiLink> <http://ru.dbpedia.org/resource/Епархия_Абаэтетубы> .`

Not sure if this answers your question.

> Have you used SERD and -f turtle to create the HDT file? All other parsers are broken.

Oh, I used the default n-triples. Maybe `-f turtle` will fix the problem.

**Comment by RubenVerborgh** [comment]
The N-Triples parser is broken (especially encoding things). SERD is in the Makefile, but might need to be switched on.

**Comment by bendiken** [comment]
> Have you used SERD and -f turtle to create the HDT file? All other parsers are broken.

@RubenVerborgh I'd take a stab this week at addressing the parser situation. Could you elaborate, please, on your experience and observations regarding the different parsers that the library (ostensibly) supports at present?

**Comment by RubenVerborgh** [comment]
There are lots of problems with the built-in parser, including incorrect interpretation of non-ASCII and/or encoded characters. My suggested course of action would be to get rid of the built-in parser and only use SERD. SERD gives spec-compatible results, and encoding is handled fine.

Only downside: direct reading of gzipped files only works with the built-in parser; piping is not supported either.

**Comment by bendiken** [comment]
@RubenVerborgh I'd also be in favor of removing the built-in parser. Serd is a high-quality library that's the right thing to do for Turtle and N-Triples support. Perhaps the gzip support could be contributed to Serd directly, benefiting a wider base of users.

Any observations regarding the Raptor integration? As HDT moves to quads, we'll need N-Quads support in hdt-cpp, and it probably makes sense to keep the optional Raptor dependency around for that, unless it's got significant problems.

**Comment by RubenVerborgh** [comment]
No experience with Raptor. Would be great if SERD supported quads as well… can't be too hard to implement.

**Comment by bendiken** [comment]
@RubenVerborgh True, I might contribute N-Quads support to Serd just so as to avoid the complicated and tangled web of Raptor's dependencies, which runs rather deep.

**Comment by RubenVerborgh** [comment]
Fixed in #31.
